### PR TITLE
Add check for invalid topic statistics publish period

### LINKED ISSUE #1147: Topic Statistics 0 publish period publishes statistics as fast as possible
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu Focal
- Installation type:
  - Fat Binary
- Version or commit hash:
  - https://ci.ros2.org/view/packaging/job/packaging_linux/1883/
- DDS implementation:
  - rmf_fastrtps_cpp
- Client library (if applicable):
  - rclcpp

#### Steps to reproduce issue

```
SubscriptionOptions options; 
options.topic_stats_options.publish_period = 0ms;
node->create_subscription(..., options);
```

#### Expected behavior

Throw an error, or don't publish, not sure which is more appropriate.

#### Actual behavior

Publishes at about 2500Hz, probably the rate of the spin loop?

### PR DESCRIPTION
**Pull request by dabonnie**
Resolves #1147 

Signed-off-by: Devin Bonnie <dbbonnie@amazon.com>

### CODE REVIEW

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 1]
```diff
@@ -81,6 +82,10 @@ create_subscription(
       options,
       *node_topics->get_node_base_interface()))
   {
+    if (options.topic_stats_options.publish_period <= std::chrono::milliseconds(0)) {
+      throw std::invalid_argument("topic statistics publish period must be greater than 0");
```
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/create_subscription.hpp:86]: i would notify the current setting as well.

File: rclcpp/include/rclcpp/subscription_options.hpp [hunk 2]
```diff
@@ -66,7 +66,7 @@ struct SubscriptionOptionsBase
     // Topic to which topic statistics get published when enabled. Defaults to /statistics.
     std::string publish_topic = "/statistics";
 
-    // Topic statistics publication period in ms. Defaults to one minute.
+    // Topic statistics publication period in ms. Defaults to one second.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/subscription_options.hpp:69]: We should document there that only values `>= 0` are allowed. And/or we should update the functions that throw to say they can throw in this situation.

File: rclcpp/include/rclcpp/create_subscription.hpp (hunk 1 above)
Comment (wjwwood) [reply rclcpp/include/rclcpp/create_subscription.hpp:86]: Seems reasonable to do this.

Comment (dabonnie) [reply rclcpp/include/rclcpp/create_subscription.hpp]: Added the current publish period to the exception message.

File: rclcpp/include/rclcpp/subscription_options.hpp (hunk 2 above)
Comment (dabonnie) [reply rclcpp/include/rclcpp/subscription_options.hpp]: Added doc

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 3]
```diff
@@ -41,10 +42,26 @@ namespace rclcpp
...
 /// Create and return a subscription of the given MessageT type.
 /**
- * The NodeT type only needs to have a method called get_node_topics_interface()
- * which returns a shared_ptr to a NodeTopicsInterface, or be a
- * NodeTopicsInterface pointer itself.
- */
+  * The NodeT type only needs to have a method called get_node_topics_interface()
+  * which returns a shared_ptr to a NodeTopicsInterface, or be a
+  * NodeTopicsInterface pointer itself.
+  * @tparam MessageT
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/create_subscription.hpp:48]: Please use `\` rather than `@`.

File: rclcpp/include/rclcpp/subscription_options.hpp [hunk 4]
```diff
@@ -66,7 +66,8 @@ struct SubscriptionOptionsBase
     // Topic to which topic statistics get published when enabled. Defaults to /statistics.
     std::string publish_topic = "/statistics";
 
-    // Topic statistics publication period in ms. Defaults to one minute.
+    // Topic statistics publication period in ms. Defaults to one second. Only values greater
+    // than zero are allowed.
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/subscription_options.hpp:70]:
```suggestion
    // Topic statistics publication period in ms. Defaults to one second.
    // Only values greater than zero are allowed.
```

File: rclcpp/include/rclcpp/create_subscription.hpp (hunk 3 above)
Comment (wjwwood) [inline rclcpp/include/rclcpp/create_subscription.hpp:48]:
```suggestion
    *
    * @tparam MessageT
```

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 5]
```diff
@@ -41,10 +42,26 @@ namespace rclcpp
...
+  * @tparam MessageMemoryStrategyT
+  * @tparam NodeT
+  * @param node
+  * @param topic_name
+  * @param qos
+  * @param callback
+  * @param options
+  * @param msg_mem_strat
+  * @return the created subscription
+  * @throws std::invalid argument if topic statistics is enabled and the publish period is
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/create_subscription.hpp:62]:
```suggestion
  * @throws std::invalid_argument if topic statistics is enabled and the publish period is
```

File: rclcpp/include/rclcpp/create_subscription.hpp (hunk 3 above)
Comment (dabonnie) [reply rclcpp/include/rclcpp/create_subscription.hpp]: My mistake (thanks auto-formatting), resolved.

File: rclcpp/include/rclcpp/subscription_options.hpp (hunk 4 above)
Comment (dabonnie) [reply rclcpp/include/rclcpp/subscription_options.hpp]: Done

**Comment by dabonnie** [comment]
@ros2/aws-oncall @mm318  - please run this CI job
Gist: https://gist.githubusercontent.com/dabonnie/9ed7799a603e9a0f04c7b0f7e4a2b42c/raw/a64b95814c4cf07f3779820bef35d8a9e2e4f137/ros2.repos
BUILD args: --packages-up-to rclcpp
TEST args: --packages-select rclcpp
Job: ci_launcher

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 6]
```diff
@@ -41,10 +42,27 @@ namespace rclcpp
...
+  * \param node
+  * \param topic_name
+  * \param qos
+  * \param callback
+  * \param options
+  * \param msg_mem_strat
+  * \return the created subscription
+  * \throws std::invalid_argument if topic statistics is enabled and the publish period is
+  * less than or equal to zero.
+  */
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/create_subscription.hpp:65]:
```suggestion
 * The NodeT type only needs to have a method called get_node_topics_interface()
 * which returns a shared_ptr to a NodeTopicsInterface, or be a
 * NodeTopicsInterface pointer itself.
 *
 * \tparam MessageT
 * \tparam CallbackT
 * \tparam AllocatorT
 * \tparam CallbackMessageT
 * \tparam SubscriptionT
 * \tparam MessageMemoryStrategyT
 * \tparam NodeT
 * \param node
 * \param topic_name
 * \param qos
 * \param callback
 * \param options
 * \param msg_mem_strat
 * \return the created subscription
 * \throws std::invalid_argument if topic statistics is enabled and the publish period is
 * less than or equal to zero.
 */
```

Comment (wjwwood) [reply rclcpp/include/rclcpp/create_subscription.hpp:65]: Should be fine to take this change without rerunning CI.

**Comment by dabonnie** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10996/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6332/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8956/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10919/)

**Comment by sloretz** [comment]
@wjwwood friendly ping, can this be merged now?

**Comment by wjwwood** [comment]
I believe so, thanks @sloretz for the reminder.

**Comment by wjwwood** [comment]
@dabonnie this landed on master now, but needs to be backported to foxy (I assume). Can you open a pull request to do that and mention me so I can review it?

**Comment by dabonnie** [comment]
> @dabonnie this landed on master now, but needs to be backported to foxy (I assume). Can you open a pull request to do that and mention me so I can review it?

Will do!

**Pull request merged**